Source/JavaScriptCore/runtime/ObjectConstructorInlines.h

```diff
@@ -63,7 +63,145 @@ ALWAYS_INLINE void objectAssignIndexedPropertiesFast(JSGlobalObject* globalObjec
+
+        if (structure->mayBePrototype()) {
+            dataLogLnIf(verbose, "target may be prototype");
+            return false;
+        }
+
+        if (structure->didPreventExtensions()) {
+            dataLogLnIf(verbose, "target has didPreventExtensions");
+            return false;
+        }
+
+        if (structure->hasBeenFlattenedBefore()) {
+            dataLogLnIf(verbose, "target has flattened before");
+            return false;
+        }
+
+        if (structure->hasBeenDictionary()) {
+            dataLogLnIf(verbose, "target has been dictionary");
+            return false;
+        }
+
+        if (structure->isBrandedStructure()) {
+            dataLogLnIf(verbose, "target has isBrandedStructure");
+            return false;
+        }
+
+        if (structure->hasPolyProto()) {
+            dataLogLnIf(verbose, "target has PolyProto");
+            return false;
+        }
+
+        if (structure->hasReadOnlyOrGetterSetterPropertiesExcludingProto()) {
+            dataLogLnIf(verbose, "target has non-writable properties");
+            return false;
+        }
+
+        if (structure->hasNonEnumerableProperties()) {
+            dataLogLnIf(verbose, "target has non-enumerable properties");
+            return false;
+        }
+
+        if (structure->hasNonConfigurableProperties()) {
+            dataLogLnIf(verbose, "target has non-configurable properties");
+            return false;
+        }
+
+        if (!structure->isQuickPropertyAccessAllowedForEnumeration()) {
+            dataLogLnIf(verbose, "target has symbol properties (right now we disable this optimization in this case since we cannot detect private symbol properties)");
+            return false;
+        }
+
+        return true;
+    };
+
+    if (!checkStrucure(targetStructure))
+        return false;
+
+    if (targetStructure->transitionWatchpointSetIsStillValid()) {
+        dataLogLnIf(verbose, "target transitionWatchpointSetIsStillValid");
+        return false;
+    }
+
+    // If the sourceStructure is frozen, we retrieve the last one before freezing.
+    if (sourceStructure->transitionKind() == TransitionKind::Freeze) {
+        dataLogLnIf(verbose, "source was frozen. Let's look into the previous structure");
+        sourceStructure = sourceStructure->previousID();
+        if (!sourceStructure)
+            return false;
+
+        dataLogLnIf(verbose, "source should have ArrayStorage since it was frozen. Let's see whether it is empty and we can quickly get the previous structure without ArrayStorage.");
+        if (sourceStructure->transitionKind() == TransitionKind::AllocateArrayStorage && !source->canHaveExistingOwnIndexedProperties()) {
+            sourceStructure = sourceStructure->previousID();
+            if (!sourceStructure)
+                return false;
+        }
+    }
+
+    if (!checkStrucure(sourceStructure))
+        return false;
+
+    if (targetStructure->inlineCapacity() != sourceStructure->inlineCapacity()) {
```

Comment:
This is fine as long as targetStructure->inlineCapacity is greater than source, right?